### DISCUSSION TITLE
Invoke cmp with specific source

### DISCUSSION THREAD

**Discussion by timsofteng** [hrsh7th/nvim-cmp#912] [General]
Hello. How can I invoke cmp manually with one or few specific sources from the insert mode?

**Comment by gegoune** [comment]
Have a look at docs, there is an example of how to do this afair.

**Comment by timsofteng** [comment]
Thanks
Find it
May I map few binds in single map command?

**Reply by gegoune** [reply]
What do you mean exactly?

**Reply by timsofteng** [reply]
I want to do invoke cmp with buffer source by pressing c-n and still have ability to circle between items by pressing same shortcuts as vim natively works.

**Reply by gegoune** [reply]
I guess you will have to come up with function mapping such that if menu is visible it selects next item and if it isn't it will open menu for you. Perhaps it's just a variation of tab mappings? Do haver a look at it and see if you can amend it to your needs.

**Reply by timsofteng** [reply]
> I guess you will have to come up with function mapping such that if menu is visible it selects next item and if it isn't it will open menu for you.

Right

**Reply by timsofteng** [reply]
How do you think? Is it possible?

**Comment by hrsh7th** [comment]
See this documents.
https://github.com/hrsh7th/nvim-cmp/blob/main/doc/cmp.txt#L188

**Comment by timsofteng** [comment]
@hrsh7th yaeh I've found it already.
Is it possible to have same map for few actions as I described below?

**Comment by hrsh7th** [comment]
@gegoune 's solution is correct.

You can use the following mapping.

```lua
cmp.setup {
  mapping = {
    ['<C-n>'] = function()
      if cmp.visible() then
        cmp.select_next_item()
      else
        cmp.complete(...)
      end
    end
  }
}
```

Basically, I try not to answer the question "How should I realize this feature?". (Because there is no end)
nvim-cmp is designed so that it can be combined as much as possible to realize any function. Please try various things.

**Comment by timsofteng** [comment]
Sorry but I'm just not good enough in lua.
I've added this into config but it doesn't work

```
    ['<C-n>'] = function()
      if cmp.visible() then
        cmp.mapping.select_next_item()
      else
        cmp.mapping.complete({
          config = {
            sources = {
              { name = 'buffer' }
            }
          }
        })
      end
    end,
    ['<C-d>'] = cmp.mapping.scroll_docs(-4),
```

**Reply by hrsh7th** [reply]
Please remove `.mapping`

**Reply by timsofteng** [reply]
Tried without `.mapping` firstly. `cmp.select_next_item()` works but `cmp.complete()` doesnt

**Comment by timsofteng** [comment]
I've finally configure it.
Thanks everyone.

@hrsh7th is it possible to change `rounded` borders to `single`?

**Reply by Shougo** [reply]
Please see this.

https://github.com/hrsh7th/nvim-cmp/pull/472

Note: It seems not documented yet?

**Reply by timsofteng** [reply]
@Shougo 
Yeah, it is absent in documantation. 
When I enable this from documentation
```
  window = {
    completion = cmp.config.window.bordered(),
    documentation = cmp.config.window.bordered(),
  },
```

The background of content and borders are transparent.
Any options to make it transparent using this manual config?
https://github.com/hrsh7th/nvim-cmp/pull/472

**Reply by hrsh7th** [reply]
https://github.com/hrsh7th/nvim-cmp/blob/main/doc/cmp.txt#L529

**Reply by timsofteng** [reply]
@hrsh7th cool!
How to make it transparent?

**Reply by hrsh7th** [reply]
What is the transparent? I can't understand.

The `window.*.border` and `window.*.winhighlight` are the same as the nvim built-in API. You can read the docs via `':help nvim_open_win`.

**Reply by timsofteng** [reply]
I mean no background
Like this
![image](https://user-images.githubusercontent.com/51912173/164458940-e3f97d0c-de6e-41d9-a78e-3fa2ca7de9ad.png)

**Reply by hrsh7th** [reply]
I don't know. You can customize `window.*.winhighlight` via yourself.

**Reply by timsofteng** [reply]
@hrsh7th 
I achive this with `FloatNormal` set to `Normal` or to `FloatNormal` which globally set in my colorscheme.

```
      winhighlight = 'Normal:Normal,FloatBorder:Normal,NormalFloat:Normal,CursorLine:Visual,Search:None',
```

**Reply by timsofteng** [reply]
Btw I've set it globally. Maybe it is make sense that `cmp` popups inherit my gobal configs?
